Purchases/Networking/RCHTTPClient.m

```diff
@@ -29,44 +32,87 @@ - (instancetype)initWithSystemInfo:(RCSystemInfo *)systemInfo {
+
+- (void)performRequest:(NSString *)httpMethod
+              serially:(BOOL)performSerially
+                  path:(NSString *)path
+                  body:(nullable NSDictionary *)requestBody
+               headers:(nullable NSDictionary<NSString *, NSString *> *)headers
+     completionHandler:(nullable RCHTTPClientResponseHandler)completionHandler {
+    if (performSerially) {
+        RCHTTPRequest *rcRequest = [[RCHTTPRequest alloc] initWithHTTPMethod:httpMethod
+                                                                        path:path
+                                                                        body:requestBody
+                                                                     headers:headers
+                                                           completionHandler:completionHandler];
+        @synchronized (self) {
+            if (self.currentSerialRequest) {
+                RCDebugLog(@"There's a request currently running and %ld requests left in the queue, queueing %@ %@",
+                           self.queuedRequests.count,
+                           httpMethod,
+                           path);
+                [self.queuedRequests addObject:rcRequest];
+                return;
+            } else {
+                RCDebugLog(@"there are no requests currently running, starting request %@ %@", httpMethod, path);
+                self.currentSerialRequest = rcRequest;
+            }
+        }
+    }
+
+    [self assertIsValidRequestWithMethod:httpMethod body:requestBody];
 
     NSMutableDictionary *defaultHeaders = self.defaultHeaders.mutableCopy;
     [defaultHeaders addEntriesFromDictionary:headers];
 
-    NSMutableURLRequest *request = [self createRequestWithMethod:HTTPMethod
-                                                            path:path
-                                                     requestBody:requestBody
-                                                         headers:defaultHeaders];
+    NSMutableURLRequest *urlRequest = [self createRequestWithMethod:httpMethod
+                                                               path:path
+                                                        requestBody:requestBody
+                                                            headers:defaultHeaders];
 
     typedef void (^SessionCompletionBlock)(NSData *_Nullable, NSURLResponse *_Nullable, NSError *_Nullable);
 
     SessionCompletionBlock block = ^void(NSData *_Nullable data,
                                          NSURLResponse *_Nullable response,
                                          NSError *_Nullable error) {
-        [self handleResponse:response data:data error:error request:request completionHandler:completionHandler];
+        [self handleResponse:response
+                        data:data
+                       error:error
+                     request:urlRequest
+           completionHandler:completionHandler
+beginNextRequestWhenFinished:performSerially];
```

Comment:
Is it possible that this block never executes if there's some sort of error when performing the request?